wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java

```diff
@@ -195,13 +206,24 @@ private Connection retryConnection(final List<HostSpec> limitlessRouters, final
       throw new SQLException(Messages.get("LimitlessConnectionPlugin.maxRetriesExceeded"), originalException);
     }
 
-    limitlessRouterService.runMonitor(this.pluginService.getHostListProvider().getClusterId(), props);
-    final List<HostSpec> newLimitlessRouters = this.limitlessRouterService.getLimitlessRouters(
-        this.pluginService.getHostListProvider().getClusterId(), props);
+    final List<HostSpec> newLimitlessRouters = synchronouslyGetLimitlessRouters(props);
     if (newLimitlessRouters == null || newLimitlessRouters.isEmpty()) {
       throw new SQLException(Messages.get("LimitlessConnectionPlugin.noRoutersAvailableForRetry"), originalException);
     }
 
     return retryConnection(newLimitlessRouters, props, retryCount, originalException);
   }
+
+  private synchronized List<HostSpec> synchronouslyGetLimitlessRouters(final Properties props) throws SQLException {
+    final List<HostSpec> currentLimitlessRouters = this.limitlessRouterService.getLimitlessRouters(
+        this.pluginService.getHostListProvider().getClusterId(), props);
+    if (!currentLimitlessRouters.isEmpty()) {
+      return currentLimitlessRouters;
+    }
+
+    LOGGER.fine(Messages.get("LimitlessConnectionPlugin.synchronouslyGetLimitlessRouters"));
+    final List<HostSpec> newLimitlessRouters = this.limitlessRouterService.forceGetLimitlessRouters(
```

Comment:
It still might be empty list in some cases. Should we add a loop?

Reply:
Good point. I'll add a while loop as well as some properties for max tries and retry intervals.